Automattic/wp-calypso
# [wip] Improve error handling

#### Changes proposed in this Pull Request

Fixes https://github.com/Automattic/wp-calypso/issues/58468

![image](https://user-images.githubusercontent.com/77539/143873878-ccf94023-9634-4e64-a3cc-ced63ab19006.png)


#### Testing instructions
There are a couple scenarios that need to be tested here, the testing flow follow the same pattern but for different scenarios we will have to modify the `wp-content/lib/atomic/eligibility.php` file in `wpcom` and sandbox the wpcom public api.
* Update `wp-content/lib/atomic/eligibility.php` to reflect the scenario you are testing.
* Navigate to `http://calypso.localhost:3000/start/woocommerce-install/confirm`
* Select your site from the site picker
* The page should render differently depending the scenario you are testing

##### Confirm screen with eligibility holds
To test this scenario we have to make sure that there is a eligibility hold present like `EXCESSIVE_DISK_SPACE` in the server response so we sandbox the api and change the eligibility endpoint to return a hold:
```php
// The value returned by get_space_used_bytes() excludes transcoded media files, but includes the original media.
	$disk_space_used = get_space_used_bytes( $blog_id );
	if ( $disk_space_used > 150 * GB_IN_BYTES) {  // make this if pass and it will return the eligibility hold
		$errors[] = build_error(
			EXCESSIVE_DISK_SPACE,
			'This site cannot be transferred due to disk space limitations.'
		);
	}
```
This should be the result:
<img width="1512" alt="Screen Shot 2021-11-30 at 11 57 11 PM" src="https://user-images.githubusercontent.com/1035546/144174533-3c51af73-35c9-4875-9020-592a5bc4dc8a.png">
* Button should be disabled
* It should show a hold alert

##### Confirm screen with eligibility warnings
To test this scenario we have to make sure that:
* The server returns eligibility warnings.
* The server doesn't return an eligibility hold, update the `eligbility.php` file so the API doesn't return a hold.

make sure that the `get_plugin_warnings()` returns all the plugin warnings to be able to test:
```php
// like this
function get_plugin_warnings() {
	$plugin_warnings = array_filter( get_unsupported_plugins(),
		function ( $plugin_element ) {
			$check_plugin_usage = __NAMESPACE__ . '\\is_using_plugin_' . $plugin_element['id'];

			return function_exists( $check_plugin_usage ) ? $check_plugin_usage() : false;
		}
	);

	// @ToDo: Remove - developing purposes. testing this 
	$plugin_warnings = array_filter( get_unsupported_plugins() );

	return array_values( $plugin_warnings );
}
```
This should be the result:
<img width="1355" alt="Screen Shot 2021-12-01 at 12 14 42 AM" src="https://user-images.githubusercontent.com/1035546/144176011-636f18c2-afcb-40e8-9764-39019fd1df37.png">
* Button should not be disabled and should link to the transfer step.
* It should show a list of the warnings.

##### Confirm screen without eligibility holds and warnings
To test this scenario we have to make sure that the endpoint doesn't return any hold or warning, this can be done by returning an empty array in the `get_plugin_warnings` and `get_widgets_warnings` functions.
```php
function get_plugin_warnings() {
	$plugin_warnings = array_filter( get_unsupported_plugins(),
		function ( $plugin_element ) {
			$check_plugin_usage = __NAMESPACE__ . '\\is_using_plugin_' . $plugin_element['id'];

			return function_exists( $check_plugin_usage ) ? $check_plugin_usage() : false;
		}
	);

	// @ToDo: Remove - developing purposes. testing this 
	$plugin_warnings = array_filter( get_unsupported_plugins() );

	return array_values([] ); // return an empty array
```
For the holds we want to make sure that the conditionals are not fired making the ifs `false` is enough (maybe there is a better way of doing this please update this comment).
This should be the result:
<img width="1354" alt="Screen Shot 2021-12-01 at 12 25 55 AM" src="https://user-images.githubusercontent.com/1035546/144176956-ae3979c6-d3dc-42f7-8c82-72d5a73a58a8.png">
* Button should be enabled
* It should only show the subdomain warning
